Add MySQLCluster v1beta2

d-kuro opened the pull request:
refs: https://github.com/cybozu-go/moco/issues/337

Add MySQLCluster v1beta2 API and conversion webhook.

## MySQLCluster v1beta2

Split the `.spec.serviceTemplate` into `.spec.primaryServiceTemplate` and `.spec.replicaServiceTemplate`.

Befor:

```yaml
spec:
  serviceTemplate:
    metadata:
      labels:
        foo: bar
```

After:

```yaml
spec:
  primaryServiceTemplate:
    metadata:
      labels:
        foo: bar
  replicaServiceTemplate:
    metadata:
      labels:
        foo: baz
```

d-kuro commented on Makefile:
```diff
@@ -20,7 +21,7 @@ SHELL = /bin/bash
 
 # Produce CRDs that work back to Kubernetes 1.11 (no version conversion)
-CRD_OPTIONS = "crd:crdVersions=v1,maxDescLen=220"
+CRD_OPTIONS = "crd:crdVersions=v1,maxDescLen=0"
```
I had to remove the description to get around the size limit.

```
Error from server (Invalid): error when creating "STDIN": CustomResourceDefinition.apiextensions.k8s.io "mysqlclusters.moco.cybozu.com" is invalid: metadata.annotations: Too long: must have at most 262144 bytes
```

d-kuro replied:
The size limit could have been circumvented with `maxDescLen=30`, but 30 characters are too small, so I set it to 0.

ymmt2005 replied:
That's unavoidable but a pity.
Hopefully, we may replace corev1.PodSpec/ServiceSpec with ApplyConfigurations in k8s.io/client-go
so that we can greatly reduce the volume of descriptions.

d-kuro replied:
I will make the changes to ApplyConfigurations in another Pull Request.

d-kuro commented on Makefile:
```diff
@@ -52,16 +53,19 @@ help: ## Display this help.
-generate: controller-gen ## Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
+generate: controller-gen conversion-gen ## Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
 	$(CONTROLLER_GEN) object:headerFile="hack/boilerplate.go.txt" paths="./..."
+	$(CONVERSION_GEN) -i ./api/v1beta2/ -o . -O zz_generated.conversion --go-header-file hack/boilerplate.go.txt
```
Use `conversion-gen` to generate the conversion code.

ymmt2005 replied:
@d-kuro 
Could you tell me if this is needed?
Because kubebuilder book does not mention this tool, I'm not sure using this is the right way or not.
https://book.kubebuilder.io/multiversion-tutorial/conversion.html

d-kuro replied:
I don't think it's necessary, but it makes sense to use conversion-gen because writing conversion code for an Object with many fields is tedious and can lead to mistakes.

conversion-gen is a proven tool that is also used in k/k and cluster-api, so I think no problem in using it.

ymmt2005 replied:
Understood.  Thank you for your explanation.

d-kuro commented on api/v1beta1/mysqlcluster_types.go:
```diff
@@ -434,6 +434,7 @@ type ReconcileInfo struct {
 
 // +kubebuilder:object:root=true
 // +kubebuilder:subresource:status
+//+kubebuilder:storageversion
```
As of this Pull Request, storageversion is still at v1beta1.

d-kuro commented on api/v1beta2/conversion.go:
```diff
@@ -0,0 +1,127 @@
+)
+
+const (
+	SpecReplicaServiceTemplateAnnotation = "mysqlcluster.v1beta2.moco.cybozu.com/spec.replicaServiceTemplate"
```
Information not present in v1beta1 will be kept in the annotation.

d-kuro commented on api/v1beta2/conversion_test.go:
```diff
@@ -0,0 +1,52 @@
+	f := roundtrip.CompatibilityTestFuzzer(scheme, nil)
+	f.NilChance(0.5).NumElements(0, 3)
+
+	t.Run("v1beta2 => v1beta1 => v1beta2", func(t *testing.T) {
```
Using fuzzing to conversion APIs to each other and test that no information is missing.

d-kuro replied:
refs: https://www.youtube.com/watch?v=HsYtMvvzDyI&t=1237s

d-kuro commented on api/v1beta2/mysqlcluster_types.go:
```diff
@@ -0,0 +1,436 @@
+	// +kubebuilder:validation:MinItems=1
+	VolumeClaimTemplates []PersistentVolumeClaim `json:"volumeClaimTemplates"`
+
+	// PrimaryServiceTemplate is a `Service` template for primary.
+	// +optional
+	PrimaryServiceTemplate *ServiceTemplate `json:"primaryServiceTemplate,omitempty"`
+
+	// ReplicaServiceTemplate is a `Service` template for replica.
+	// +optional
+	ReplicaServiceTemplate *ServiceTemplate `json:"replicaServiceTemplate,omitempty"`
```
This is the only change from v1beta1.

d-kuro commented on charts/moco/CHANGELOG.md:
```diff
@@ -5,6 +5,46 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 
+### Support MySQLCluster v1beta2 API
```
Manual operations are required to support conversion webhooks in Helm Chart, which is documented in the changelog.

ymmt2005 commented on api/v1beta2/mysqlcluster_webhook.go:
```diff
@@ -0,0 +1,11 @@
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
```
I think the defaulting and validating webhooks should be implemented for `v1beta2`.

d-kuro replied:
I thought this was unnecessary in v1beta2, because if conversion is working, the validating/mutating webhook defined in v1beta1 will do the job.

ymmt2005 replied:
@d-kuro 
Have you experimented with that area?
That's only true when the webhook configuration `matchPolicy` is set to `true`.
https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-matchpolicy

I'd strongly suggest implementing the webhook for v1beta2 because v1beta1 will soon be removed.

d-kuro replied:
Thanks for the helpful links!
I checked the operation locally, but I didn't have enough knowledge that it wasn't because the conversion was running 🙏 

As pointed out in another comment, I'm going to move the conversion code to v1beta1, so I will implement it by the way.

d-kuro replied:
I added AdmissionWebhook to v1beta2 API.
https://github.com/cybozu-go/moco/pull/350/commits/42da61056edfa67d7a91dfc15129c0ad3d306e7c

ymmt2005 commented on Makefile:
```diff
@@ -52,16 +53,19 @@ help: ## Display this help.
 manifests: controller-gen kustomize ## Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
 	$(CONTROLLER_GEN) $(CRD_OPTIONS) rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
-	$(KUSTOMIZE) build config/kustomize-to-helm/overlays/crds -o charts/moco/crds
+	$(KUSTOMIZE) build config/crd -o config/crd/tests
```
Is `config/crd/tests` a directory for temporary files?
If so, please ignore this directory by `.gitignore`.

d-kuro replied:
I wanted to enable conversion webhooks in controller tests, so I used manifests with a patch for testing.
https://github.com/cybozu-go/moco/blob/693d9edb720b30e7d012c3319a71d66a8b1963d1/controllers/suite_test.go#L67

But, I have rewritten all controller tests to use v1beta2, so it may not be necessary now.
I will try to clean up the code.

d-kuro replied:
At the time of this pull request, the storage version was v1beta1, so I needed to run the conversion webhook on controller-test as well.

d-kuro replied:
added comments: https://github.com/cybozu-go/moco/pull/350/commits/b22d8b0174df63c8a7c193c24836318e78ec49bb

d-kuro commented on api/v1beta1/conversion.go:
```diff
@@ -0,0 +1,4 @@
+package v1beta1
+
+// Hub marks this type as a conversion hub.
+func (*MySQLCluster) Hub() {}
```
I see, I will fix it.

d-kuro replied:
Moved conversion logic to v1beta1:
https://github.com/cybozu-go/moco/pull/350/commits/a94e04a2b00f6d88a7062c54af51bec1989b39cb

d-kuro commented:
😢 

> The following packages have unmet dependencies:
> mysql-client : Depends: mysql-client-8.0 but it is not going to be installed

refs: https://github.com/actions/virtual-environments/pull/4674

zoetrope commented:
When I ran `make generate`, I got the following errors.
It seems that you need to rename the conversion functions.

```
❯ make generate
/home/ikezoe/go/src/github.com/cybozu-go/moco/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
/home/ikezoe/go/src/github.com/cybozu-go/moco/bin/conversion-gen -i ./api/v1beta1/ -o . -O zz_generated.conversion --go-header-file hack/boilerplate.go.txt
E1217 15:08:21.047602  453076 conversion.go:193] Rename function github.com/cybozu-go/moco/api/v1beta1 Convert__MySQLClusterSpec_To_v1beta2_MySQLClusterSpec -> Convert_v1beta1_MySQLClusterSpec_To_v1beta2_MySQLClusterSpec to match expected conversion signature     
E1217 15:08:21.047803  453076 conversion.go:193] Rename function github.com/cybozu-go/moco/api/v1beta1 Convert__MySQLCluster_To_v1beta2_MySQLCluster -> Convert_v1beta1_MySQLCluster_To_v1beta2_MySQLCluster to match expected conversion signature
E1217 15:08:21.047930  453076 conversion.go:193] Rename function github.com/cybozu-go/moco/api/v1beta1 Convert_v1beta2_MySQLClusterSpec_To__MySQLClusterSpec -> Convert_v1beta2_MySQLClusterSpec_To_v1beta1_MySQLClusterSpec to match expected conversion signature     
E1217 15:08:21.048078  453076 conversion.go:193] Rename function github.com/cybozu-go/moco/api/v1beta1 Convert_v1beta2_MySQLCluster_To__MySQLCluster -> Convert_v1beta2_MySQLCluster_To_v1beta1_MySQLCluster to match expected conversion signature
E1217 15:08:21.070427  453076 conversion.go:755] Warning: could not find nor generate a final Conversion function for github.com/cybozu-go/moco/api/v1beta1.MySQLClusterSpec -> github.com/cybozu-go/moco/api/v1beta2.MySQLClusterSpec
E1217 15:08:21.070477  453076 conversion.go:756]   the following fields need manual conversion:
E1217 15:08:21.070483  453076 conversion.go:758]       - ServiceTemplate
E1217 15:08:21.071268  453076 conversion.go:755] Warning: could not find nor generate a final Conversion function for github.com/cybozu-go/moco/api/v1beta2.MySQLClusterSpec -> github.com/cybozu-go/moco/api/v1beta1.MySQLClusterSpec
E1217 15:08:21.071309  453076 conversion.go:756]   the following fields need manual conversion:
E1217 15:08:21.071317  453076 conversion.go:758]       - PrimaryServiceTemplate
E1217 15:08:21.071323  453076 conversion.go:758]       - ReplicaServiceTemplate
```

zoetrope commented:
> The following packages have unmet dependencies:
> mysql-client : Depends: mysql-client-8.0 but it is not going to be installed

`mysql-client` may be pre-installed.
How about stopping `mysql-client` from being installed?

d-kuro commented on api/v1beta1/mysqlcluster_types.go:
```diff
@@ -434,6 +434,7 @@ type ReconcileInfo struct {
 
 // +kubebuilder:object:root=true
 // +kubebuilder:subresource:status
+//+kubebuilder:storageversion
```
Thanks 🙏 
https://github.com/cybozu-go/moco/pull/350/commits/63c1df5b9d3855f9670248013ee65268bb68b61d

d-kuro commented:
@zoetrope 

> When I ran make generate, I got the following errors.
> It seems that you need to rename the conversion functions.

Sorry, I fixed conversion codes.
It is working fine now.
https://github.com/cybozu-go/moco/runs/4602212152?check_suite_focus=true

ymmt2005 commented on api/v1beta1/backuppolicy_types.go:
```diff
@@ -33,7 +34,7 @@ type BackupPolicySpec struct {
 
 	// Specifies parameters for backup Pod.
-	JobConfig JobConfig `json:"jobConfig"`
+	JobConfig mocov1beta2.JobConfig `json:"jobConfig"`
```
Using v1beta2 types to define v1beta1 types seems like a bug to me.

d-kuro replied:
I did this because BackupPolicy exists in the v1beta1 package and I wanted it to use a common struct with MySQLCluster in v1beta2.
Fixed to use after conversion.
https://github.com/cybozu-go/moco/pull/350/commits/4f5c0ae1595d2b591397a17b7fd8ac68133421cc

ymmt2005 commented on api/v1beta1/groupversion_info.go:
```diff
@@ -17,4 +17,6 @@ var (
 	// AddToScheme adds the types in this group-version to the given scheme.
 	AddToScheme = SchemeBuilder.AddToScheme
+
+	localSchemeBuilder = &SchemeBuilder.SchemeBuilder
```
Is this used?  there seem no references to this.

d-kuro replied:
Used in the code generated by conversion-gen.
I added comments.
https://github.com/cybozu-go/moco/pull/350/commits/e6e44e68758bac34671ce8720857c796de73ccf9

https://github.com/cybozu-go/moco/blob/4f5c0ae1595d2b591397a17b7fd8ac68133421cc/api/v1beta1/zz_generated.conversion.go#L20

ymmt2005 commented on api/v1beta1/mysqlcluster_webhook.go:
```diff
@@ -22,8 +22,6 @@ func (r *MySQLCluster) SetupWebhookWithManager(mgr ctrl.Manager) error {
 
 // EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
 
-//+kubebuilder:webhook:path=/mutate-moco-cybozu-com-v1beta1-mysqlcluster,mutating=true,failurePolicy=fail,sideEffects=None,groups=moco.cybozu.com,resources=mysqlclusters,verbs=create,versions=v1beta1,name=mmysqlcluster.kb.io,admissionReviewVersions=v1
```
Why don't we remove this file altogether?

d-kuro replied:
Thanks！, I remove it: https://github.com/cybozu-go/moco/pull/350/commits/c7988e3dda9a4053a9a79f0ccd19cc62b5a762e0

But, a conversion webhook is necessary, so we left the webhook registration process.

ymmt2005 commented on cmd/moco-controller/cmd/run.go:
```diff
@@ -113,6 +115,11 @@ func subMain(ns, addr string, port int) error {
+		return err
+	}
+
 	if err = (&mocov1beta1.BackupPolicy{}).SetupWebhookWithManager(mgr); err != nil {
```
With `matchPolicy=Eqiuvalent`, the webhook for v1beta2 can handle v1beta1 so this is not needed.

d-kuro replied:
I left the conversion webhook logic in v1beta1 because it is still there.

ymmt2005 commented on controllers/mysqlcluster_controller.go:
```diff
@@ -871,8 +871,14 @@ func (r *MySQLClusterReconciler) reconcileV1BackupJob(ctx context.Context, req c
 			VolumeSource: *jc.WorkVolume.DeepCopy(),
 		}}
 
+		var bucketConfig mocov1beta2.BucketConfig
+
+		if err := mocov1beta1.Convert__BucketConfig_To_v1beta2_BucketConfig(&jc.BucketConfig, &bucketConfig, nil); err != nil {
+			return fmt.Errorf("failed to convert bucket config from v1beta1 to v1beta2: %w", err)
+		}
```
We should have retrieved BackupPolicy with v1beta2 at line 841 so that
this conversion can be removed.
https://github.com/cybozu-go/moco/pull/350/commits/4f5c0ae1595d2b591397a17b7fd8ac68133421cc#diff-67f34c9b4cabb4950b383eab55b6ebeeebc524fa70c888bdc132662c16a7899fR841

d-kuro replied:
I assumed that I would only create the v1beta2 API for `MySQLCluster` and continue to use v1beta1 without upgrading the `BackupPolicy`.
This conversion is necessary because `MySQLCluster` (v1beta2 package) and `BackupPolicy` (v1beta1 package) use a common `JobConfig` structure.

Or should I create a v1beta2 API for BackupPolicy too, even if there are no changes to the API?

ymmt2005 replied:
@d-kuro 
> Or should I create a v1beta2 API for BackupPolicy too, even if there are no changes to the API?

I think so.

@zoetrope Thoughts?

zoetrope replied:
I think the same version would be better.

d-kuro replied:
I added `v1beta2.BackupPolicy`

* Add BackupPolicy v1beta2: https://github.com/cybozu-go/moco/pull/350/commits/d0fa6af7719539826eacb0b46bcb87714462534b
* Use BackupPolicy v1beta2: https://github.com/cybozu-go/moco/pull/350/commits/7c02683f30de1d502d84cdd40fb4b51d1616fe8c
* Add conversion webhook for BackupPolicy: https://github.com/cybozu-go/moco/pull/350/commits/4cc80be617fe401251a76ab0f1e9c9f1b7d9985d

ymmt2005 reviewed (commented):
@d-kuro 
It looks almost good to me.  Thank you.

Please update apiVersion in examples and documents.
I believe this is the last one.

zoetrope commented:
@d-kuro 
Reviewed. And I confirmed that conversion webhook works.
It's fine.

I have one question, why did you change test package from `v1beta1` to `v1beta1_test`?

d-kuro commented:
@ymmt2005 
> Please update apiVersion in examples and documents.
> I believe this is the last one.

Thanks! I added BackupPolicy v1beta2 docs.
https://github.com/cybozu-go/moco/pull/350/commits/e2400d9421c30873645c5c5161bf05a30fdccb81

d-kuro commented:
@zoetrope 

> I have one question, why did you change test package from v1beta1 to v1beta1_test?

Thank you comments.
I changed the package to avoid circular imports, but I realized it wasn't necessary now.
Fixed: https://github.com/cybozu-go/moco/pull/350/commits/05700b91abc6cdb4b5e1e0bcc46c54ddf1795d42

ymmt2005 commented:
@d-kuro I mean, please update apiVersion in examples and documents.
For example, https://github.com/cybozu-go/moco/blob/d-kuro/conversion/examples/anti-affinity.yaml#L2

ymmt2005 reviewed (changes requested):
Documents and examples hasn't been updated yet.

d-kuro commented:
@ymmt2005 
Sorry, I must have missed that.
I fixed it:
https://github.com/cybozu-go/moco/pull/350/commits/fcbc2a567274400f7df2d8926412cb908692864c
https://github.com/cybozu-go/moco/pull/350/commits/6625785d52f07bb3afef0861a5c44e3a6a816a2b

ymmt2005 commented on docs/reconcile.md:
```diff
@@ -83,7 +83,9 @@ MOCO creates three Services for each MySQLCluster, that is:
-The Services' labels, annotations, and `spec` fields can be customized with MySQLCluster's `status.serviceTemplate` field.
+The Services' labels, annotations, and `spec` fields can be customized with MySQLCluster's `spec.primaryServiceTemplate` and `spec.replicaServiceTemplate` field.
+The `spec.primaryServiceTemplate` configures the Service for the primary mysqld instance
+and the `spec.replicaServiceTemplate` configures the Service for the replica mysql instance.
```
```suggestion
and the `spec.replicaServiceTemplate` configures the Service for the replica mysqld instances.
```

ymmt2005 reviewed (approved):
LGTM

zoetrope reviewed (approved):
LGTM
